locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsByonLocationResolver.java

```diff
@@ -61,129 +56,140 @@
+        if (hosts instanceof String) {
+            if (((String) hosts).isEmpty()) {
+                hostIds = ImmutableList.of();
+            } else {
+                hostIds = WildcardGlobs.getGlobsAfterBraceExpansion("{"+hosts+"}",
+                        true /* numeric */, /* no quote support though */ PhraseTreatment.NOT_A_SPECIAL_CHAR, PhraseTreatment.NOT_A_SPECIAL_CHAR);
+            }
+        } else if (hosts instanceof Iterable) {
+            hostIds = ImmutableList.copyOf((Iterable<?>)hosts);
+        } else {
+            throw new IllegalArgumentException("Invalid location '"+spec+"'; at least one host must be defined");
+        }
+        if (hostIds.isEmpty()) {
+            throw new IllegalArgumentException("Invalid location '"+spec+"'; at least one host must be defined");
+        }
 
-        List<String> hostIdentifiers = WildcardGlobs.getGlobsAfterBraceExpansion("{"+hosts+"}",
-                true /* numeric */, /* no quote support though */ PhraseTreatment.NOT_A_SPECIAL_CHAR, PhraseTreatment.NOT_A_SPECIAL_CHAR);
-        List<JcloudsSshMachineLocation> machines = Lists.newArrayList();
-        
-        for (String hostIdentifier : hostIdentifiers) {
-            Map<?, ?> machineFlags = MutableMap.builder()
-                    .put("id", hostIdentifier)
-                    .putIfNotNull("user", user)
-                    .putIfNotNull("privateKeyFile", privateKeyFile)
-                    .build();
-            try {
-                // TODO management of these machines may be odd, as it is passed in as a key as config to a spec
-                JcloudsSshMachineLocation machine = jcloudsLocation.rebindMachine(jcloudsLocation.config().getBag().putAll(machineFlags));
-                machines.add(machine);
-            } catch (NoMachinesAvailableException e) {
-                log.warn("Error rebinding to jclouds machine "+hostIdentifier+" in "+jcloudsLocation, e);
-                Exceptions.propagate(e);
+        final List<Map<?,?>> machinesFlags = Lists.newArrayList();
+        for (Object hostId : hostIds) {
+            Map<?, ?> machineFlags;
+            if (hostId instanceof String) {
+                machineFlags = parseMachineFlags((String)hostId, config);
+            } else if (hostId instanceof Map) {
+                machineFlags = parseMachineFlags((Map<String,?>)hostId, config);
+            } else {
+                throw new IllegalArgumentException("Invalid host type '"+(hostId == null ? null : hostId.getClass().getName())+", referrenced in spec '"+spec);
             }
+            machinesFlags.add(machineFlags);
         }
         
+        Supplier<List<JcloudsMachineLocation>> machinesFactory = new Supplier<List<JcloudsMachineLocation>>() {
+            @Override
+            public List<JcloudsMachineLocation> get() {
+                // FIXME Can't change the parent of the JcloudsMachineLocation; but the FixedListMachineLocation will expect it to be the parent?
```

Comment:
Unfortunately not, and unfortunately I can't find the log where we encountered this! I tried to reproduce in a simple unit test but didn't manage. Will need to investigate further.

You're right that a vague FIXME comment in the code is not the right thing. I'll remove the fixme, and try to reproduce properly so I can open a jira issue.